[pytorch] Fix pushLong() issue in pickler.

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28057 [pytorch] Fix pushLong() issue in pickler.**

For pushLong() in Pickler, it looks like we only use for a single use case, with a 10-byte value.

We were handling > 256 bytes incorrectly, by using a LONG4 opcode (expecting 4-byte length), but pushing 8 bytes. We could harden this handling, but rather than improve codepaths that we never expect to use, this change simply removes the incorrect codepath and adds and assert.

Differential Revision: [D17934174](https://our.internmc.facebook.com/intern/diff/D17934174/)

Commit:
[pytorch] Fix LONG4 length in pickler.

LONG4 length should be a 4-byte integer, per:
  https://github.com/python/cpython/blob/2.7/Lib/pickletools.py#L677

Document also seems to indicate that it should be signed.
Also, support on the unpickle side.

Differential Revision: [D17934174](https://our.internmc.facebook.com/intern/diff/D17934174/)

[ghstack-poisoned]

zdevito reviewed (changes requested):
LONG4 is suppose to first read 4 bytes to determine the size of the data to read and then read that many bytes. I am not sure what our implementation does (either before or after the patch). LONG1 looks more correct, though even that only supports numbers that are precisely 8 bytes. @driazati for what is suppose to happen here. I think we should just delete LONG4 handling, and assert in LONG1 handling that length == 8 on write (currently we only do it on read).

jjlilley commented:
Thanks for the feedback.                                                                                                                                         
I'm obviously not the expert in pickle opcodes, mostly noticed this in the other change and wanted to avoid an incorrect-size read.                              
AFAICT, it looks like there's only one caller of the pushLong function, and it's called with a 10-byte magic number, so asserting size==8 on writing a LONG1 isn\
't going to universally work.

driazati commented:
`LONG1` of size 10 is needed for #25109, but I don't think we have a real use case for `LONG4` in the pickler, since TorchScript's ints are all `int64_t`s. I think we should just remove the `LONG4` handling and error out if the code tries to go down that path.

jjlilley commented:
In any case, closing this specific PR.

Commit:
Update on "[pytorch] Fix LONG4 length in pickler."


LONG4 length should be a 4-byte integer, per:
  https://github.com/python/cpython/blob/2.7/Lib/pickletools.py#L677

Document also seems to indicate that it should be signed.
Also, support on the unpickle side.

Differential Revision: [D17934174](https://our.internmc.facebook.com/intern/diff/D17934174/)

[ghstack-poisoned]

Commit:
Update on "[pytorch] Fix pushLong() issue in pickler."


For pushLong() in Pickler, it looks like we only use for a single use case, with a 10-byte value.

We were handling > 256 bytes incorrectly, by using a LONG4 opcode (expecting 4-byte length), but pushing 8 bytes. We could harden this handling, but rather than improve codepaths that we never expect to use, this change simply removes the incorrect codepath and adds and assert.

Differential Revision: [D17934174](https://our.internmc.facebook.com/intern/diff/D17934174/)

[ghstack-poisoned]

jjlilley commented:
> `LONG1` of size 10 is needed for #25109, but I don't think we have a real use case for `LONG4` in the pickler, since TorchScript's ints are all `int64_t`s. I think we should just remove the `LONG4` handling and error out if the code tries to go down that path.

That sounds reasonable, uploading a version that simply removes the incorrect codepath and asserts that the size fits in 1 byte.

driazati commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -415,17 +415,11 @@ void Pickler::pushDouble(double value) {
-  }
+  TORCH_INTERNAL_ASSERT(
+    size <= std::numeric_limits<uint8_t>::max(),
+    "Cannot pickle a long with a size larger than 256 bytes");
```
This should say `Cannot pickle a long with a size larger than 1 byte`

jjlilley replied:
I can change this message, but given that data.size() can be up to 256, saying that the long's size can't be larger than 1 byte doesn't seem accurate.

I could say the more-awkward-but-accurate:
    "Cannot pickle a long with the size of size is larger than 1 byte"

Or, is there less-awkward-but-also-accurate preferred version of the assert text?

driazati replied:
Maybe just `Cannot pickle a long larger than 255 bytes` (0 is a valid size, 256 is not)

jjlilley replied:
fixed, thanks!

Commit:
Update on "[pytorch] Fix pushLong() issue in pickler."


For pushLong() in Pickler, it looks like we only use for a single use case, with a 10-byte value.

We were handling > 256 bytes incorrectly, by using a LONG4 opcode (expecting 4-byte length), but pushing 8 bytes. We could harden this handling, but rather than improve codepaths that we never expect to use, this change simply removes the incorrect codepath and adds and assert.

Differential Revision: [D17934174](https://our.internmc.facebook.com/intern/diff/D17934174/)

[ghstack-poisoned]
